File: go/terminal/ui/dialog.go
package ui

import "github.com/nullentropy/caution/go/terminal/gfx"

var dialogTitleFont = gfx.NewFont(15, gfx.FontOpts{Weight: 600})

// dialogTitleH is the card's title band - the dialog.titleHeight token.
func dialogTitleH() float32 { return Metrics.DialogTitleH }

// Dialog is a modal: a full-viewport scrim that swallows input, with a
// centered card whose children lay out (frame/anchors) against the content
// area below the title. Z-order comes from tree order - the server appends
// dialogs last. Clicking the scrim or pressing Escape emits `dismiss`; the
// server decides whether to remove the node.
type Dialog struct {
	Core
	Title     string
	CardW     float32
	CardH     float32
	OnDismiss func()
}

func NewDialog() *Dialog {
	d := &Dialog{CardW: 420, CardH: 200}
	d.self = d
	return d
}

func (d *Dialog) Interactive() bool { return true } // the scrim swallows every miss

func (d *Dialog) cardRect() gfx.Rect {
	b := d.Bounds
	return gfx.R(b.X+(b.W-d.CardW)/2, b.Y+(b.H-d.CardH)/2, d.CardW, d.CardH)
}

func (d *Dialog) LayoutChildren() {
	d.adopt()
	card := d.cardRect()
	d.LayoutInto(gfx.R(card.X, card.Y+dialogTitleH(), card.W, card.H-dialogTitleH()))
}

func (d *Dialog) PaintSelf(dl *gfx.DisplayList) {
	dl.Fill(d.Bounds, gfx.WithAlpha(gfx.Black, 0.45), gfx.Corners{})
	card := d.cardRect()
	dl.Shadow(card, gfx.WithAlpha(gfx.Black, 0.5), 40, gfx.CornerRadius(Metrics.RadiusPanel), 0, 18)
	dl.Rect(card, *Theme["panel"], gfx.RectOpts{
		Radius: gfx.CornerRadius(Metrics.RadiusPanel), BorderWidth: Metrics.BorderWidth, BorderColor: Theme["edge"],
	})
	m := d.UI.Measure(dialogTitleFont, d.Title)
	dl.Text(d.Title, card.X+20, card.Y+(dialogTitleH()-(m.Ascent+m.Descent))/2, dialogTitleFont, *Theme["ink"])
	hair := Metrics.BorderWidth
	dl.Fill(gfx.R(card.X, card.Y+dialogTitleH()-hair, card.W, hair), *Theme["edge"], gfx.Corners{})
}

func (d *Dialog) OnPointerDown(x, y float32, _ int) {
	if !Contains(d.cardRect(), x, y) && d.OnDismiss != nil {
		d.OnDismiss()
	}
}

File: src/ui/dialog.ts
import { BLACK, withAlpha } from '../gfx/color';
import { font } from '../gfx/font';
import { rect, Rect } from '../gfx/geom';
import type { DisplayList } from '../gfx/painter';
import { metrics } from './metrics';
import type { SemanticSpec } from './semantics';
import { theme } from './theme';
import { contains, Widget } from './widget';

const TITLE_FONT = font(15, { weight: 600 });
/** the card's title band, the dialog.titleHeight token */
const titleH = (): number => metrics.dialogTitleHeight;

/**
 * Modal dialog: a full-viewport scrim that swallows input, with a centered
 * card whose children lay out (frame/anchors) against the content area below
 * the title. Z-order comes from tree order and the server appends dialogs last.
 * Clicking the scrim or pressing Escape emits `dismiss`. The server decides
 * whether to remove the node.
 */
export class Dialog extends Widget {
  title = '';
  cardW = 420;
  cardH = 200;
  onDismiss: (() => void) | null = null;

  override get interactive(): boolean {
    return true; // the scrim swallows every click that misses the card
  }

  cardRect(): Rect {
    const b = this.bounds;
    return rect(b.x + (b.w - this.cardW) / 2, b.y + (b.h - this.cardH) / 2, this.cardW, this.cardH);
  }

  override layoutChildren(): void {
    this.adopt();
    const card = this.cardRect();
    this.layoutChildrenInto(rect(card.x, card.y + titleH(), card.w, card.h - titleH()));
  }

  override paintSelf(dl: DisplayList): void {
    dl.rect(this.bounds, { color: withAlpha(BLACK, 0.45) });
    const card = this.cardRect();
    dl.shadow(card, { radius: metrics.radiusPanel, blur: 40, offset: [0, 18], color: withAlpha(BLACK, 0.5) });
    dl.rect(card, { color: theme.panel, radius: metrics.radiusPanel, borderWidth: metrics.borderWidth, borderColor: theme.edge });
    const m = this.ui!.measure(TITLE_FONT, this.title);
    dl.text(this.title, card.x + 20, card.y + (titleH() - (m.ascent + m.descent)) / 2, TITLE_FONT, theme.ink);
    const hair = metrics.borderWidth;
    dl.rect(rect(card.x, card.y + titleH() - hair, card.w, hair), { color: theme.edge });
  }

  override onPointerDown(x: number, y: number): void {
    if (!contains(this.cardRect(), x, y) && this.onDismiss) {
      this.sound('close');
      this.onDismiss();
    }
  }

  override semantics(): SemanticSpec {
    return { role: 'text', label: `dialog: ${this.title}` };
  }
}
